server/api.go

```diff
@@ -32,6 +32,22 @@ type apiServer struct {
 	*Server
 }
 
+// enforce authorization policy per action/subject/object
+func (a *apiServer) enforcePolicy(subject string, object string, action string) (bool, error) {
+	// Load policy data
+	// [NOTE] casbin raise a panic if it fails to load the policy, i.e: policy file is corrupted,
+	// Refer to issue: https://github.com/casbin/casbin/issues/640
+	// [TODO] Policy storage and modification should be improved later
+	err := a.authzEnforcer.LoadPolicy()
```

Comment:
To be honest, I am not sure. In this particular case, Casbin loads data from on disk csv file. Without having in depth knowledge about Casbin, I can only guess that this operation should be implemented in an efficient way.

However, since it is anyway an I/O operation, it will introduce overhead. How much overheat is it ? I think we have to measure. A local execution would not be accurate. Do you have any suggestion ?

I traced to this discussion https://github.com/casbin/casbin/issues/45 , which mentioned performance overhead on large environment.

However, apart from loading the policy in memory on each request, I wonder if there is any other solution to keep policy fresh in memory .

Reply:
To be sure, I can:

- Set up a small K8s cluste
- Deploy a few node on it
- Create about a dozen of streams
- Measure latency in clients' responses in case of Authz enabled and another case for authz disabled.
But I am not even sure that this measurement can be considered accurate enough. Any thoughts ?

Reply:
My main concern here is that, as I understand, `LoadPolicy()` is going to disk on _every_ Liftbridge RPC to reload the Casbin policy. This seems inefficient since, _generally_, I would expect Casbin policy updates to be relatively infrequent.

Here is my proposition: what if instead of reloading the policy every time, we require the operator to signal a reload? This allows them to modify the on-disk policy files and then, when ready to apply, signal the Liftbridge server, e.g. by sending a `SIGHUP` signal. `SIGHUP` is commonly used by systems to force a reload of configuration, e.g. see [Nginx](https://docs.nginx.com/nginx/admin-guide/basic-functionality/runtime-control/#controlling-nginx) and [NATS](https://docs.nats.io/running-a-nats-service/nats_admin/signals#reload-server-configuration).

I think a longer term goal for Liftbridge will be to support this type of "hot reloading" for most of its server configuration, so once that is implemented, `SIGHUP` could be used to reload _all_ server configuration, including authz and the server's yaml configuration.

What do you think?

Reply:
This is actually a great idea. I have not thought about that before.  New things learned every day.

I will make a simple version to handle `SIGHUP` and add a `sync.RWMutex` lock to permission policy. I will keep it simple this time because:
- Eventually this mechanism can be replaced by permission sync (via Raft probably or may be something else)

- For configuration, we will have a separate MR for that. However, it seems that [Viper](https://github.com/spf13/viper#watching-and-re-reading-config-files) actually has a concept called config watching.

Thanks for the tips.

Reply:
implemented